//@ edition:2021

macro_rules! demo2 {
    ( $a:tt $b:tt ) => { println!("two tokens") };
}

macro_rules! demo3 {
    ( $a:tt $b:tt $c:tt ) => { println!("three tokens") };
}

macro_rules! demo4 {
    ( $a:tt $b:tt $c:tt $d:tt ) => { println!("four tokens") };
}

fn main() {
    // Non-ascii identifiers
    demo2!(Ã‘"foo"); //~ ERROR prefix `Ã‘` is unknown
    demo4!(Ã‘#""#);  //~ ERROR prefix `Ã‘` is unknown
    demo3!(ðŸ™ƒ#"");
    //~^ ERROR identifiers cannot contain emoji
}
